tools/api_boost/api_boost.py

```diff
@@ -0,0 +1,153 @@
+    # and is a hack that will go away when we remove v2.
+    if re.match(API_INCLUDE_REGEX, line) and 'envoy/service/auth/v2alpha' not in line:
+      continue
+    output_lines.append(line)
+
+  # Write to temporary file. We can't overwrite in place as we're executing
+  # concurrently with other ApiBoostFile() invocations that might need the file
+  # we're writing to.
+  pathlib.Path(path + TMP_SWP_SUFFIX).write_text('\n'.join(output_lines) + '\n')
+
+
+# Replace the original file with the temporary file created by ApiBoostFile()
+# for a given path.
+def SwapTmpFile(path):
+  pathlib.Path(path + TMP_SWP_SUFFIX).rename(path)
+
+
+# Update the Envoy source tree the latest API.
+def ApiBoostTree(args):
+  # Optional setup of state. We need the API type database, compilation database
+  # and api_booster tool in place before we can start boosting.
+  if args.generate_api_type_database:
+    sp.run(['./tools/gen_api_type_database.sh'], check=True)
+
+  if args.generate_compilation_database:
+    sp.run(['./tools/gen_compilation_database.py', '--run_bazel_build', '--include_headers'],
+           check=True)
+
+  if args.build_api_booster:
+    # Similar to gen_compilation_database.py, we only need the cc_library for
+    # setup. The long term fix for this is in
+    # https://github.com/bazelbuild/bazel/issues/9578.
+    dep_build_targets = [
+        '//source/...',
+        '//test/...',
+    ]
+    query = 'kind(cc_library, {})'.format(' union '.join(dep_build_targets))
+    dep_lib_build_targets = sp.check_output(['bazel', 'query', query]).decode().splitlines()
+    sp.run([
+        'bazel',
+        'build',
+        '--strip=always',
+        '//tools/clang_tools/api_booster',
+    ] + dep_lib_build_targets,
+           check=True)
+
+  # Figure out where the LLVM include path is. We need to provide this
+  # explicitly as the api_booster is built inside the Bazel cache and doesn't
+  # know about this path.
+  # TODO(htuch): this is fragile and depends on Clang version, should figure out
+  # a cleaner approach.
+  llvm_include_path = os.path.join(
+      sp.check_output([os.getenv('LLVM_CONFIG'), '--libdir']).decode().rstrip(),
+      'clang/9.0.0/include')
+
+  # Determine the files in the target dirs eligible for API boosting.
+  paths = []
+  for p in args.paths:
+    if os.path.isfile(p):
+      paths.append(p)
+      continue
+    for root, dirs, files in os.walk(p):
+      for f in files:
+        if f.endswith('.cc') or f.endswith('.h'):
+          paths.append(os.path.join(root, f))
+
+  # Filter files that aren't in the compilation database; we can't meaningfully
+  # run them under Clang libtooling (e.g. Windows plaform files on Linux).
+  known_files = set(
+      [entry['file'] for entry in json.loads(pathlib.Path('compile_commands.json').read_text())])
+  paths = [p for p in paths if p in known_files]
+
+  # The API boosting is file local, so this is trivially parallelizable, use
```

Comment:
At higher level I think you can actually use aspect to parallelize this by Bazel, and it should be very similar to what compile db aspect does.

Reply:
It's an interesting idea, but I feel it's going to be a lot more complicated for not a big gain. As we see in https://github.com/envoyproxy/envoy/issues/9266, there is a lot of complexity involved in mapping from the paths that show up under aspects (in Bazel cache) and physical source paths. Given that this will eventually be rewriting the source almost in place, I think it will be easier to reason about and debug if we avoid adding Bazel machinery.

Reply:
There are ways to do inline rewrite within `bazel run` (just like gazelle does), and it actually reduces the complexity of mapping source to bazel cache. I'm fine to keep this for now and eventually I can move this to aspect.

Reply:
Yeah, if we have some straightforward examples to crib from here, I'm good to switch  to that pattern or leave it to later. I'm mostly worried about being sucked into a customer Bazel engineering exercise here.

Reply:
As discussed offline, we'll go with `multiprocessing` for now and consider cleanup later. Since this tool is manually invoked by developers only at major version upgrade time (once per year), we can tolerate some rough edges.